middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml

```diff
@@ -0,0 +1,431 @@
+          P.define_global_symbol (Symbol.linkage_name_as_string function_symbol)
+          @ acc
+      in
+      (* We build here the **reverse** list of fields for the function slot *)
+      match closure_code_pointers with
+      | Full_application_only ->
+        let acc =
+          P.int ~dbg closure_info
+          :: P.symbol_from_linkage_name ~dbg code_linkage_name
+          :: acc
+        in
+        acc, slot_offset + 2, env, Ece.pure, updates
+      | Full_and_partial_application ->
+        let acc =
+          P.symbol_from_linkage_name ~dbg code_linkage_name
+          :: P.int ~dbg closure_info
+          :: P.symbol_from_linkage_name ~dbg
+               (Linkage_name.create (C.curry_function_sym arity))
+          :: acc
+        in
+        acc, slot_offset + 3, env, Ece.pure, updates)
+
+  let rec fill_layout ~set_of_closures_symbol_ref symbs decls dbg ~startenv
+      value_slots env effs acc updates ~starting_offset slots =
+    match slots with
+    | [] -> List.rev acc, starting_offset, env, effs, updates
+    | (slot_offset, slot) :: slots ->
+      let acc =
+        if starting_offset > slot_offset
+        then
+          Misc.fatal_errorf "Starting offset %d is past slot offset %d"
+            starting_offset slot_offset
+        else if starting_offset = slot_offset
+        then acc
+        else
+          acc
+          @ List.init (slot_offset - starting_offset) (fun _ -> P.int ~dbg 1n)
+      in
+      let acc, next_offset, env, eff, updates =
+        fill_slot ~set_of_closures_symbol_ref symbs decls dbg ~startenv
+          value_slots env acc ~slot_offset updates slot
+      in
+      let effs = Ece.join eff effs in
+      fill_layout ~set_of_closures_symbol_ref symbs decls dbg ~startenv
+        value_slots env effs acc updates ~starting_offset:next_offset slots
+end
+
+(* Filling-up of dynamically-allocated sets of closures. *)
+module Dynamic = Make_layout_filler (struct
+  type cmm_term = Cmm.expression
+
+  let int ~dbg i = C.nativeint ~dbg i
+
+  let simple env simple =
+    let term, env, eff = C.simple env simple in
+    `Data [term], env, eff
+
+  let infix_header ~dbg ~first_value_slot_offset =
+    C.alloc_infix_header first_value_slot_offset dbg
+
+  let symbol_from_linkage_name ~dbg linkage_name =
+    C.symbol_from_linkage_name ~dbg linkage_name
+
+  let define_global_symbol _ = []
+end)
+
+(* Filling-up of statically-allocated sets of closures. *)
+module Static = Make_layout_filler (struct
+  type cmm_term = Cmm.data_item
+
+  let int ~dbg:_ i = C.cint i
+
+  let simple env simple =
+    let env, contents = C.simple_static env simple in
+    contents, env, Ece.pure
+
+  let infix_header ~dbg:_ ~first_value_slot_offset =
+    C.cint (C.infix_header first_value_slot_offset)
+
+  let symbol_from_linkage_name ~dbg:_ linkage_name =
+    C.symbol_address (Linkage_name.to_string linkage_name)
+
+  let define_global_symbol sym = List.rev (C.define_symbol ~global:true sym)
```

Comment:
I'd prefer that this function did not perform the `rev`, and instead the caller of this function (in the functor), performed the rev. This makes it much simpler to reason, and would allow to use `List.rev_append` which is more efficient (or at the very least, this function should be renamed with a `_rev` somewhere in its name).

Reply:
Fixed